### PR TITLE
crimson/os: support multicore seastore

### PR DESCRIPTION
**Pull request by liu-chunmei**
new implement segments assign and management for each shard. 
next will support  shard zns_sm_metadata_t for each shard.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by liu-chunmei** [comment]
this PR include #48575, since it is based on #48575.

**Comment by liu-chunmei** [comment]
jenkins test make check

**Comment by athanatos** [comment]
@liu-chunmei Sounds like you got some feedback from @cyx1231st already?  Should I review this version or wait for the next version?

**Comment by cyx1231st** [comment]
@athanatos I just gave the initial direction that static disk partition should be managed at the general `Device` level, in preparation for the next step: dynamic partitioning across shards. The other parts should be kept the same in the sharded SeaStore implementation.

@liu-chunmei is this version ready for review according to the above design, or is there anything different in the current design?

**Comment by liu-chunmei** [comment]
> @liu-chunmei Sounds like you got some feedback from @cyx1231st already? Should I review this version or wait for the next version?

@athanatos cyx1231 review cyanstore PR, this seastore PR include cyanstore PR. so you can review seastore part.  I will work on cyanstore comments and update it.

**Comment by liu-chunmei** [comment]
> @athanatos I just gave the initial direction that static disk partition should be managed at the general `Device` level, in preparation for the next step: dynamic partitioning across shards. The other parts should be kept the same in the sharded SeaStore implementation.
> 
> @liu-chunmei is this version ready for review according to the above design, or is there anything different in the current design?

@cyx1231st what do you mean Device level?  the static segments assignment is done in segment_manager.  I think for different manager should have different implementation.

**Comment by cyx1231st** [comment]
> @cyx1231st what do you mean Device level? the static segments assignment is done in segment_manager. I think for different manager should have different implementation.

Right, different manager may have different underlying implementations, but **all devices** must include the same/similar feature (may not implement in this PR though), and be managed in the same way from higher level SeaStore perspective. I suppose the boundray of the abstraction is at the general `Device` class.

File: src/crimson/os/seastore/seastore.h
Comment (cyx1231st) [inline src/crimson/os/seastore/seastore.h:506]: Move it to the private section.

Please also scan the code to move other methods to the private section if it is not supposed to be called by OSD.

File: src/crimson/os/seastore/seastore.cc
Comment (cyx1231st) [inline src/crimson/os/seastore/seastore.cc:1913]: ceph_assert(seastar::this_shard_id() == PRIMARY_CORE);

Please also scan the code to add this assert to other public interfaces that should only be available in PRIMARY_CORE.

File: src/test/crimson/seastore/transaction_manager_test_state.h
Comment (cyx1231st) [inline src/test/crimson/seastore/transaction_manager_test_state.h:54]: Looks we can merge `_start()` and `_init()`.

File: src/crimson/os/seastore/segment_manager/block.cc
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_manager/block.cc:577]: I think the overall mkfs logic can be restructured with less repeating code, for example for each device:

`SeaStore::mkfs()` on the primary core only, call `Device::mkfs()`, mostly to (maybe) create the device and to write down the initial superblock.

`SeaStore::ShardStores::mkfs()` on each core (including the primary core), call `Device::mkfs_sharded()`, to include the common mkfs logic in each device shard.

**Review by cyx1231st — COMMENTED** [review]
Looks working!

These are still high-level reviews about major suggestions. Please also rebase the code.

File: src/crimson/os/seastore/seastore.h
Comment (liu-chunmei) [reply src/crimson/os/seastore/seastore.h:141]: @athanatos currently osd doesn't know how seastore implement shard(they are decoupled), and from one store instance to access store shard.  Yingxin suggested each shardservice in OSD should get one shardstore instance(shardstore instance exposed to OSD), and needn't do an interface transform, so what do you think?

Comment (cyx1231st) [reply src/crimson/os/seastore/seastore.h:141]: As suggested in https://github.com/ceph/ceph/pull/48575#discussion_r1052815989, I think this is an invariant at the `FuturizedStore` level that also applies to the sharded SeaStore.

Comment (liu-chunmei) [reply src/crimson/os/seastore/seastore.h:141]: > @athanatos currently osd doesn't know how seastore implement shard(they are decoupled), and from one store instance to access store shard. Yingxin suggested each shardservice in OSD should get one shardstore instance(shardstore instance exposed to OSD), and needn't do an interface transform, so what do you think?

@athanatos Could you please take a look on this interface implementation?

File: src/crimson/os/seastore/segment_manager.h
Comment (cyx1231st) [reply src/crimson/os/seastore/segment_manager.h:27]: cc @athanatos this is what I was mentioned in the standup.

**Comment by liu-chunmei** [comment]
@cyx1231st please review first, I will rebase later since the code changed so fast.

**Comment by cyx1231st** [comment]
@liu-chunmei please proceed to rebase and test. Ping me if there are anything uncertain around the previous comments.

**Comment by liu-chunmei** [comment]
jenkins test make check

**Comment by liu-chunmei** [comment]
jenkins test make check

File: src/crimson/os/futurized_store.h
Comment (cyx1231st) [inline src/crimson/os/futurized_store.h:70]: How about dropping open_collection/create_new_collection/do_transaction from FuturizedStore and use them through get_local_shard_store()?

File: src/crimson/os/futurized_store.h
Comment (cyx1231st) [inline src/crimson/os/futurized_store.h:68]: See above, who is the caller of FuturizedStore::read() ?

File: src/crimson/os/futurized_store.h
Comment (cyx1231st) [inline src/crimson/os/futurized_store.h:29]: How about deducing the primary core of FuturizedStore in its constructors, thus relax this constraint at the level of futurized_store?

File: src/crimson/os/futurized_store.h
Comment (cyx1231st) [inline src/crimson/os/futurized_store.h:33]: The empty line and the comment are unnecessary.

File: src/crimson/os/futurized_store.h
Comment (cyx1231st) [inline src/crimson/os/futurized_store.h:87]: Unnecessary empty line.

File: src/crimson/os/futurized_store.h
Comment (cyx1231st) [inline src/crimson/os/futurized_store.h:98]: This comment is unnecessary.

File: src/crimson/os/futurized_store.h
Comment (liu-chunmei) [reply src/crimson/os/futurized_store.h:70]: I think it is better keep its transform within futurized store, don't expose it to osd. otherwise each osd parts calling these interfaces need know it call shard store.

File: src/crimson/os/futurized_store.h
Comment (liu-chunmei) [reply src/crimson/os/futurized_store.h:68]: osdmeta will call store->read.

File: src/crimson/os/futurized_store.h
Comment (cyx1231st) [reply src/crimson/os/futurized_store.h:70]: Looks to me that all these duplicated interfaces (open_collection/create_new_collection/do_transaction/read) are all related to the meta-coll operations. Proper cleanups with OSDMeta to work with FuturizedShardStore can address the repeating redirections. Also it looks reasonable to access the local FuturizedShardStore instead for the meta-coll operations.

The point is that it is simpler/more-straiforward to make FuturizedStore/FuturizedShardStore interfaces disjoint.

File: src/crimson/os/futurized_store.h
Comment (cyx1231st) [reply src/crimson/os/futurized_store.h:68]: see above

File: src/crimson/os/futurized_store.h
Comment (liu-chunmei) [reply src/crimson/os/futurized_store.h]: done in a new commit.

File: src/crimson/os/futurized_store.h
Comment (liu-chunmei) [reply src/crimson/os/futurized_store.h]: done in a new commit.

File: src/crimson/os/futurized_store.h
Comment (cyx1231st) [reply src/crimson/os/futurized_store.h:86]: I meant to drop mount()/umount(), because it looks there's no user from OSD, and it's better to expose them from `FuturizedStore` only as global operations.

File: src/crimson/os/futurized_store.h
Comment (cyx1231st) [reply src/crimson/os/futurized_store.h:84]: see above

File: src/crimson/os/futurized_store.h
Comment (cyx1231st) [inline src/crimson/os/futurized_store.h:151]: `list_collections()` is supposed to be a global interface across shards, move it to `FuturizedStore`.

It's only user is `PGShardManager::load_pgs()` called in `OSD::start()`, which is also a part of OSD global initialization.

Let's adjust it to `PGShardManager::load_pgs(FuturizedStore&)` to call this from `FuturizedStore` instead.

File: src/crimson/os/futurized_store.cc
Comment (cyx1231st) [inline src/crimson/os/futurized_store.cc:25]: Drop the future from FuturizedStore::create()'s return value.

File: src/crimson/os/alienstore/alien_store.h
Comment (cyx1231st) [inline src/crimson/os/alienstore/alien_store.h:177]: It is not needed, there should be no internal changes to AlienStore.

The AlienStore can inherit from both FuturizedStore and FuturizedShardStore, with no special internal changes to work exactly as before. This is also why the interfaces from FuturizedStore and FuturizedShardStore should be disjoint.

File: src/crimson/os/cyanstore/cyan_store.h
Comment (cyx1231st) [inline src/crimson/os/cyanstore/cyan_store.h:109]: Create another public section to group non-overriding public interfaces.

File: src/test/crimson/seastore/transaction_manager_test_state.h
Comment (cyx1231st) [inline src/test/crimson/seastore/transaction_manager_test_state.h:269]: Seems not needed.

File: src/crimson/os/futurized_store.h
Comment (cyx1231st) [inline src/crimson/os/futurized_store.h:64]: nit, rename to get_sharded_store.

File: src/crimson/tools/store_nbd/fs_driver.cc
Comment (cyx1231st) [inline src/crimson/tools/store_nbd/fs_driver.cc:148]: create a member `FuturizedShardStore &sharded_fs`

File: src/crimson/osd/osd_meta.cc
Comment (cyx1231st) [inline src/crimson/osd/osd_meta.cc:29]: Switch `OSDMeta::store` to use `FuturizedShardStore`

File: src/crimson/osd/osd.cc
Comment (cyx1231st) [inline src/crimson/osd/osd.cc:161]: Worth a TODO:
// TODO: encapsulate into OSDMeta to use its FuturizedShardStore

Or fix it directly in this PR if convenient.

File: src/crimson/osd/osd.cc
Comment (cyx1231st) [inline src/crimson/osd/osd.cc:171]: Worth a TODO:
// TODO: encapsulate into OSDMeta to use its FuturizedShardStore

Or fix it directly in this PR if convenient.

File: src/crimson/osd/osd.cc
Comment (cyx1231st) [inline src/crimson/osd/osd.cc:272]: Worth a TODO:
// TODO: encapsulate into OSDMeta to use its FuturizedShardStore

Or fix it directly in this PR if convenient.

**Review by cyx1231st — COMMENTED** [review]
Still looking, mostly about the FuturizedStore.

File: src/crimson/os/seastore/seastore.h
Comment (cyx1231st) [inline src/crimson/os/seastore/seastore.h:583]: Looks root/device/max_object_size/is_test are no longer necessary in `SeaStore` with the above suggestions.

File: src/crimson/os/futurized_store.h
Comment (liu-chunmei) [reply src/crimson/os/futurized_store.h:86]: FuturizedStore will call FuturizedShardStore mount/umount,  so what do you mean drop them?

Comment (cyx1231st) [reply src/crimson/os/futurized_store.h:86]: Why cannot drop the mount/umount interfaces from FuturizedShardStore? I don't see there are any external users (e.g. OSD), or am I missing anything?

Comment (cyx1231st) [reply src/crimson/os/futurized_store.h:86]: And I don't see FuturizedStore is calling mount/umount at the FuturizedShardStore level.

Comment (liu-chunmei) [reply src/crimson/os/futurized_store.h:86]: I understand what you mean. will remove them from FuturizedShardStore.  only define them in SeaShardStore and CyanShardStore.

File: src/crimson/os/futurized_store.cc
Comment (liu-chunmei) [reply src/crimson/os/futurized_store.cc:24]: const ConfigValues& values  alienstore will use it in start(), so put path and values as start() input parameters.

File: src/test/crimson/seastore/transaction_manager_test_state.h
Comment (liu-chunmei) [reply src/test/crimson/seastore/transaction_manager_test_state.h:269]: need call shard_stores.stop() in _destroy(), so it should return seastar::future<>.

Comment (cyx1231st) [reply src/test/crimson/seastore/transaction_manager_test_state.h:269]: I meant in the parent implementation, `EphemeralTestState::_destroy()` is already called after `EphemeralTestState::_teardown()`, right?

Comment (liu-chunmei) [reply src/test/crimson/seastore/transaction_manager_test_state.h]: right, fix it.

File: src/crimson/os/seastore/device.h
Comment (cyx1231st) [inline src/crimson/os/seastore/device.h:90]: I think the mkfs paths in SeaStore is still too complicated.

Like `FuturizedStore` vs `FuturizedShardStore`, I think it is reasonable to split into `Device` and `ShardedDevice`, so that `SeaStore` only needs to worry about global device initializations like `Device::mkfs()`, `mount()`, `start()` and `stop()`, and after that, `SeaShardStore` can get its own `ShardedDevice` through `Device::get_sharded_device()`.

There are other reasons to split the `Device` class, for example, the global member like `superblock` should be operated by the `BlockSegmentManager` rather than `ShardedBlockSegmentManager`. And I suppose there will be more implementations in `BlockSegmentManager` to support cross-shard dynamic partition in the future.

**Review by cyx1231st — COMMENTED** [review]
The logic above `SeaStore` looks good to me. Probably need another look from @athanatos for the `FuturizedStore` and `FuturizedShardStore` interface changes.

The code structure below `SeaStore` seems still need iterations to become cleaner, see my initial comments below.

File: src/crimson/os/futurized_store.h
Comment (athanatos) [inline src/crimson/os/futurized_store.h:64]: Hmm, seems like this should probably take a core_id_t?

Comment (athanatos) [reply src/crimson/os/futurized_store.h:64]: Or does it have to be called on the shard on which it will be used?  Please add a comment in the standard interface style explaining the rules.

File: src/crimson/os/futurized_store.h
Comment (athanatos) [inline src/crimson/os/futurized_store.h:73]: const

File: src/crimson/os/futurized_store.h
Comment (athanatos) [inline src/crimson/os/futurized_store.h:76]: This is perhaps a minor point, but FuturizedStore->FuturizedShardStore is confusing and redundant.  If we nest this within FuturizedStore, we can simply name it Shard or FuturizedStore::Shard.  This would then lead to CyanStore::Shard, AlienStore::Shard, and Seastore::Shard.  At the very least, I'd prefer FuturizedStoreShard to FuturizedShardStore (ditto for the other three).

File: src/crimson/os/futurized_store.h
Comment (athanatos) [inline src/crimson/os/futurized_store.h:29]: The division of methods between FuturizedStore and FuturizedShardStore makes sense to me.

File: src/crimson/os/seastore/device.h
Comment (athanatos) [reply src/crimson/os/seastore/device.h:90]: It's really not clear what the intended distinction between mkfs and shard_mkfs is here.

File: src/crimson/os/seastore/seastore.h
Comment (athanatos) [inline src/crimson/os/seastore/seastore.h:191]: This needs an interface comment, I'm not sure what the intent is.

File: src/crimson/os/seastore/seastore.h
Comment (athanatos) [inline src/crimson/os/seastore/seastore.h:197]: This needs an interface comment, I'm not sure what the intent is.

File: src/crimson/os/seastore/seastore.h
Comment (athanatos) [inline src/crimson/os/seastore/seastore.h:202]: This needs an interface comment, I'm not sure what the intent is.

**Review by athanatos — CHANGES_REQUESTED** [review]
The FuturizedStore/FuturizedShardStore method division makes sense to me.  My main feedback otherwise is that the newly added methods in seastore particularly in the mkfs pathway need standard interface comments explaining how they are meant to be used.

File: src/crimson/os/seastore/device.h
Comment (cyx1231st) [reply src/crimson/os/seastore/device.h:90]: Seems to me that mkfs() and shard_mkfs() are similar to CyanStore::mkfs() and CyanShardStore::mkfs(), that mkfs() is to mkfs at device-level (e.g. superblock), and shard_mkfs() is to mkfs at shard-level (e.g. shard-tracker).

This is why I suggest to hide the entire mkfs process inside the device, so that seastore mkfs for the device will look much simpler.

**Review by cyx1231st — COMMENTED** [review]
LGTM except for the specific suggestions.

**Comment by liu-chunmei** [comment]
jenkins test api

**Comment by liu-chunmei** [comment]
jenkins test make check

**Comment by liu-chunmei** [comment]
jenkins test api

File: src/crimson/os/seastore/onode_manager/staged-fltree/node.h
Comment (liu-chunmei) [reply src/crimson/os/seastore/onode_manager/staged-fltree/node.h:155]: this static cause heap-used-after-free bug, since every get_invalid() will return the same address, and it is freed in Cursor constructor, if next get_invalid() called after that free, it will return a null pointer.  I meet some times but not often before, this times easily trigger that bug, so fix it.

Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/node.h:155]: Worth to create a dedicated commit for an independent fix

**Review by cyx1231st — APPROVED** [review]
LGTM! Need a second approval for merge.

**Review by athanatos — APPROVED** [review]
LGTM!

**Comment by athanatos** [comment]
@liu-chunmei Needs a rebase

**Comment by liu-chunmei** [comment]
CI test: seems with this PR get the same result with main branch except for a download error. 
main branch without this PR: https://pulpito.ceph.com/yingxin-2023-04-19_05:15:13-crimson-rados-wip-chunmei-multicore-os-crimson-only-distro-default-smithi/
with this PR:  https://pulpito.ceph.com/yingxin-2023-04-20_03:37:46-crimson-rados-wip-chunmei-multicore-os-crimson-only-distro-default-smithi/

**Comment by athanatos** [comment]
@liu-chunmei Seems fine to merge once the conflict is resolved.

**Pull request merged**
